src/Solvers/matrix_solver_utils.jl

```diff
@@ -0,0 +1,236 @@
+
+@inline validate_laplacian_size(N, dim) = dim == true ? N : 1
+  
+@inline ensure_diagonal_elements_are_present!(A) = fkeep!(A, (i, j, x) -> (i == j || !iszero(x)))
+
+"""
+Precondition the solver on the GPU() with a direct LU (or Choleski) type 
+of preconditioner would require too much computation for the ldiv!(P, r) step
+and completely hinder the performances
+
+Therefore, the choice of preconditioners are limited to
+
+on the CPU
+Identity() (no preconditioner)
+ilu() (superior to everything)
+sparse_inverse_preconditioner() (not performant on CPU)
+
+on the GPU
+Identity() (no preconditioner)
+sparse_inverse_preconditioner() (sparse approximate inverse)
+
+The preconditioner settings are
+
+Identity()                    -> none 
+ilu                           -> τ = drop tolerance (the lower the more elements allowed)
+sparse_inverse_preconditioner -> ε = residual tolerance (the lower the closer to the actual inverse)
+                              -> nzrel = number of maximum elements allowed in the column / number of A element in the column
+
+"""
+
+@inline arch_preconditioner(::Val{false}, args...)                       = Identity()
+@inline arch_preconditioner(::Val{true}, ::CPU, A, τ = 0.001)            = ilu(A, τ = τ) 
+@inline arch_preconditioner(::Val{true}, ::GPU, A, ε = 0.1, nzrel = 2.0) = sparse_inverse_preconditioner(A, ε = ε, nzrel = nzrel) 
+
+@inline architecture(::CuSparseMatrixCSC) = GPU()
+@inline architecture(::SparseMatrixCSC)   = CPU()
+
+struct JacobiPreconditioner{V}
+    invdiag::V
+end
+
+# Constructor:
+function JacobiPreconditioner(A::AbstractMatrix)
+    M    = size(A, 1)
+    arch = architecture(A)
+    dev  = device(arch)
+    
+    invdiag = arch_array(arch, zeros(eltype(A), M))
+    col, row, val = unpack_constructors(arch, constructors(arch, A))
+
+    loop! = _get_inv_diag!(dev, 256, M)
+    event = loop!(invdiag, col, row, val; dependencies=Event(dev))
+    wait(dev, event)
+
+    return JacobiPreconditioner(invdiag)
+end
+
+@kernel function _multiply_in_place!(u, invdiag, v)
+    i = @index(Global, Linear)
+    @inbounds begin
+        u[i] = invdiag[i] * v[i]
+    end
+end  
+
+function  LinearAlgebra.ldiv!(u, precon::JacobiPreconditioner, v)
+    invdiag = precon.invdiag
+    arch = architecture(invdiag)
+    dev  = device(arch)
+    
+    M = length(invdiag)
+    loop! = _multiply_in_place!(dev, 256, M)
+    event = loop!(u, invdiag, v; dependencies=Event(dev))
+    wait(dev, event)
+end
+
+function  LinearAlgebra.ldiv!(precon::JacobiPreconditioner, v)
+    ldiv!(v,precon,v)
+end
+
+abstract type AbstractInversePreconditioner{M} end
+
+function  LinearAlgebra.ldiv!(u, precon::AbstractInversePreconditioner, v)
+    mul!(u, matrix(precon), v)
+end
+
+function  LinearAlgebra.ldiv!(precon::AbstractInversePreconditioner, v)
+    mul!(v, matrix(precon), v)
+end
+
+struct MitgcmPreconditioner{M} <: AbstractInversePreconditioner{M}
+    Minv :: M
+end
+
+function mit_gcm_preconditioner(A::AbstractMatrix)
```

Comment:
@simone-silvestri does this do

Equation (35) in https://agupubs.onlinelibrary.wiley.com/doi/epdf/10.1029/96JC02775 

it looks like it may currently be 1/(main diagonal) - so you could call it InvDiagPreconditioner for now, until equation (35) is coded up (we can do that later).

Reply:
Since it actually was the same type of `SparseInversePreconditioner` I unified the types and change the name to the function to `simplified_inverse_preconditioner`. @christophernhill it is trying to do `M = -A/diag(A)` on the off-diagonals and `diag(M) = 1/diag(A)`. I am not sure it is correct though, I haven't tested it much. if we want we can check it together